File: 0279-perfect-squares/0279-perfect-squares.js
/**
 * @param {number} n
 * @return {number}
 */
var numSquares = function(n) {
  const squares = [];
  let b = 1;
  while (b * b <= n) {
    squares.push(b * b);
    b++;
  }
  const dp = new Array(n + 1).fill(Infinity);
  dp[0] = 0;
  for (let i = 1; i <= n; i++) {
    for (const sq of squares) {
      if (i >= sq) dp[i] = Math.min(dp[i], dp[i - sq] + 1);
    }
  }
  return dp[n];
};

File: 0279-perfect-squares/0279-perfect-squares.rb
# @param {Integer} n
# @return {Integer}
def num_squares(n)
  squares, b = [], 1
  while b*b <= n
    squares << b*b
    b += 1
  end
  dp = [1/0.0]*(n+1)
  dp[0] = 0
  (1..n).each do |i|
    squares.each do |sq|
      dp[i] = [dp[i], dp[i-sq]+1].min if i >= sq
    end
  end
  dp[n]
end